src/V3Width.cpp

```diff
@@ -6520,6 +6520,61 @@ class WidthVisitor final : public VNVisitor {
         }
     }
 
+    bool areAggregatesComparable(AstNodeDType* lhs, AstNodeDType* rhs) {
+        if (!lhs || !rhs) return false;
+
+        lhs = lhs->skipRefp();
+        rhs = rhs->skipRefp();
+
+        if (VN_IS(lhs, UnpackArrayDType) && VN_IS(rhs, UnpackArrayDType)) {
+            auto* lhsp = VN_CAST(lhs, UnpackArrayDType);
+            auto* rhsp = VN_CAST(rhs, UnpackArrayDType);
+
+            // Compare element types
+            if (!areAggregatesComparable(lhsp->subDTypep(), rhsp->subDTypep())) return false;
+
+            // Compare size
+            std::vector<AstUnpackArrayDType*> lhsDims = lhsp->unpackDimensions();
+            std::vector<AstUnpackArrayDType*> rhsDims = rhsp->unpackDimensions();
+            if (lhsDims.size() != rhsDims.size()) return false;
+            for (size_t i = 0; i < lhsDims.size(); ++i) {
+                int lsz = lhsDims[i]->elementsConst();
+                int rsz = rhsDims[i]->elementsConst();
+                if (lsz >= 0 && rsz >= 0 && lsz != rsz) return false;
+            }
```

Comment:
Can we just use similarDType?  For that matter can much of this whole function use similarDType?

If not, please make a AstUnpackArrayDType function that does this comparison directly so we don't make temporary vectors just to check equality.

Reply:
Thanks for the insight. Yeah, I can refactor most of the logic to use similarDType. But for unpacked arrays, it ends up calling the base class (`AstArrayDType`)'s similarDType, which is stricter than what I actually need.

for example,

```systemverilog
reg [3:0] array_2 [2:0];
reg [3:0] array_3 [3:1];
```

`similarDType` returns false because the ranges don't match exactly, but it's still a valid comparison since the size count is the same.

Should I just change `AstArrayDType::similarDTypeNode` to loosen it up, or make a separate similarDTypeNode just for `AstUnpackArrayDType` with relaxed rules? just feels off if the base class is actually doing a stricter check than the subclass.

Reply:
We don't want to change `similarDType` as then they will be considered the same type and get collapsed.  So I guess make a similar function.